### DISCUSSION TITLE
Tanstack table customize expandable rows

### DISCUSSION THREAD

**Discussion by clarkkamorr24** [TanStack/table#5367] [General]
is there any solutions how can I customize the subrows like this?

![casino](https://github.com/TanStack/table/assets/43429968/0b950f4c-9387-4f1a-afcb-5563f8cb6146)

**Comment by ronkristoff** [comment]
I also want to do this but I cannot find examples using tanstack query. I found a similar behavior using https://www.mantine-react-table.com/docs/guides/detail-panel

**Reply by KevinVandy** [reply]
well, "mantine-react-table" was built with TanStack Table, so your example is right there.
"Lazy Detail Panel Fetching" is also pretty strait-forward. https://www.material-react-table.com/docs/examples/lazy-detail-panel

**Comment by andredewaard** [comment]
I think i found a workaround for this although you cant really leverage the columns api of tanstack/table
Example is made in Vue though so you need to change it to React.
```vue
          <template v-if="table.getRowModel().rows?.length">
            <template v-for="row in table.getRowModel().rows" :key="row.id">
              <TableRow :data-state="row.getIsSelected() && 'selected'">
                <TableCell v-for="cell in row.getVisibleCells()" :key="cell.id">
                  <Skeleton v-if="isLoading" class="h-5" />
                  <FlexRender
                    v-else
                    :render="cell.column.columnDef.cell"
                    :props="cell.getContext()"
                  />
                </TableCell>
              </TableRow>
              <TableRow v-if="row.getIsExpanded()">
                <TableCell :colspan="row.getVisibleCells().length">
                  // Render custom view
                </TableCell>
              </TableRow>
            </template>
          </template>
```


```ts
export const columns = (t: TFunction): ColumnDef<ClientCollectionResource, string>[] => [
  columnHelper.display({
    id: 'select',
    header: ({ table }) => h(Checkbox, {
      'checked': table.getIsAllPageRowsSelected(),
      'onUpdate:checked': value => table.toggleAllPageRowsSelected(!!value),
      'ariaLabel': 'Select all',
    }),
    cell: ({ row }) => {
      return h(Checkbox, {
        'checked': row.getIsSelected(),
        'onUpdate:checked': value => row.toggleSelected(!!value),
        'ariaLabel': 'Select row',
      })
    },
    enableSorting: false,
    enableHiding: false,
  }),
  columnHelper.accessor('id', {
    header: ({ column }) => h(DataTableColumnHeader, { column, title: t('clients.table.header.id') }),
    cell: (info) => h('div', { class: 'font-medium' }, info.getValue())
  }),
  columnHelper.accessor('name', {
    header: ({ column }) => h(DataTableColumnHeader, { column, title: t('clients.table.header.name') }),
    cell: (info) => h('div', { class: 'text-left font-medium' }, info.getValue()),
  }),
  columnHelper.display({
    id: 'actions',
    size: 1,
    cell: ({ row }) => {
      // const client = row.original
      return h('button', { class: 'flex justify-end', onClick: () => row.toggleExpanded() }, 'toggle')
    }
  })
]

const table = useVueTable({
  get data() {
    return data
  },
  get columns() {
    return columns
  },
})
```

<img width="1382" alt="Screenshot 2024-03-15 at 11 10 37" src="https://github.com/TanStack/table/assets/7560637/4d6d0cb1-7449-4523-bcb4-6e08747b4601">

I hope there would be a more native way to do this though. so we could add this custom view to the columns array somehow

**Comment by benmcginnis** [comment]
I am also looking for something similar, @andredewaard would you mind posting the markup that your example generates?  I have some strict wcag requirements and I would want the custom sub-row to have an appropriate col-span attribute, not sure how best to do that.  

I'm interested in a design like the flowbite expandable table block: https://flowbite.com/blocks/application/advanced-tables/

**Reply by andredewaard** [reply]
@benmcginnis 
This is the output you get, i don't think this will meet the wcag requirements but at least you could've something. 
Will try to find a better solution though.
<img width="759" alt="Screenshot 2024-04-29 at 13 23 13" src="https://github.com/TanStack/table/assets/7560637/245401e9-afaf-4a55-bf13-defadd2c9d16">
